### PR TITLE
[ISSUE #3473] Optimize the problem of delayed message delivery duplication

### PR DESCRIPTION
**Pull request by Git-Yang**
## What is the purpose of the change
- Optimize the problem of delayed message delivery duplication.

## Brief changelog

#3473 

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

**Comment by HScarb** [comment]
persist delay message offset table after shutting down, LGTM

File: store/src/main/java/org/apache/rocketmq/store/schedule/ScheduleMessageService.java [hunk 1]
```diff
@@ -317,7 +318,7 @@ public void executeOnTimeup() {
                         long nextOffset = offset;
                         int i = 0;
                         ConsumeQueueExt.CqExtUnit cqExtUnit = new ConsumeQueueExt.CqExtUnit();
-                        for (; i < bufferCQ.getSize(); i += ConsumeQueue.CQ_STORE_UNIT_SIZE) {
+                        for (; i < bufferCQ.getSize() && isStarted(); i += ConsumeQueue.CQ_STORE_UNIT_SIZE) {
```
Comment (HScarb) [inline store/src/main/java/org/apache/rocketmq/store/schedule/ScheduleMessageService.java:321]: `executeOnTimeup()` method will only invoke in `run()` method, which already has the condition of `isStarted()`, so why add a `isStarted()` here
https://github.com/apache/rocketmq/pull/3478/files#diff-350d772aa64ef5e5f5012863595bc2e9d35c731fbba75a81b823cf00b1a2d593R281-R282

Comment (Git-Yang) [reply store/src/main/java/org/apache/rocketmq/store/schedule/ScheduleMessageService.java:321]: When we close the service, we cannot guarantee that it has been executed in this for loop.

**Comment by Git-Yang** [comment]
1. After calling shutdown, wait for the execution of the thread to complete. At this time, it is forbidden to submit a new thread.
2. First updateOffset and then execute schedule to avoid the inability to execute updateOffset after shutdown is called

File: store/src/main/java/org/apache/rocketmq/store/schedule/ScheduleMessageService.java (hunk 1 above)
Comment (HScarb) [reply store/src/main/java/org/apache/rocketmq/store/schedule/ScheduleMessageService.java:321]: I see

**Review by HScarb — APPROVED** [review]

**Comment by Git-Yang** [comment]
### Test Results
- Before optimization
   - First step
      - 1w messages are sent through the script, and each message generates a unique self-increasing Key (0~9999)
      - After the message is sent, the delayed message is still being delivered, and the broker service is restarted at this time.
   - Second step
      - Detect whether the message is lost or repeated according to the unique self-increment key.
     - The test result is that no messages are lost and there are 7395 message duplications.
image
image
- After optimization
   - Repeat the above steps
   - The test result is no data loss, no data duplication.
image
image
   - Strengthen the testing process
      - 100w messages are sent through the script, and each message generates a unique self-increasing Key (0~999999)
      - The message delivery process restarts the service multiple times.
      - The test result is no data loss, no data duplication.
image
image

**Comment by XiaoyiPeng** [comment]
Looks Good To Me.

**Comment by MatrixHB** [comment]
> 1. After calling shutdown, wait for the execution of the thread to complete. At this time, it is forbidden to submit a new thread.
> 2. First updateOffset and then execute schedule to avoid the inability to execute updateOffset after shutdown is called

When shutdown is called, will the method `deliverExecutorService.schedule()` throw exception？？

**Review by MatrixHB — APPROVED** [review]

**Comment by Git-Yang** [comment]
> > 1. After calling shutdown, wait for the execution of the thread to complete. At this time, it is forbidden to submit a new thread.
> > 2. First updateOffset and then execute schedule to avoid the inability to execute updateOffset after shutdown is called
> 
> When shutdown is called, will the method `deliverExecutorService.schedule()` throw exception？？

Right

**Comment by Git-Yang** [comment]
Included in PR https://github.com/apache/rocketmq/pull/3478 , merged.

**Pull request closed without merging**